File: backend/app/core/permissions.py
from __future__ import annotations

from app.core.enums import Permission, UserRole

USER_PERMISSIONS = frozenset(
    {
        Permission.SEARCH_USE,
        Permission.RAG_USE,
        Permission.PROFILE_MANAGE,
        Permission.HISTORY_MANAGE,
        Permission.SAVED_MANAGE,
    }
)
EDITOR_PERMISSIONS = USER_PERMISSIONS | {
    Permission.EDITOR_ACCESS,
    Permission.MANAGED_DOCUMENTS_VIEW,
    Permission.DOCUMENT_METADATA_EDIT,
    Permission.DOCUMENT_STATUS_CHANGE,
    Permission.DOCUMENT_REINDEX,
    Permission.EDITOR_JOBS_VIEW,
}
PERMISSION_MATRIX: dict[UserRole, frozenset[Permission]] = {
    UserRole.USER: USER_PERMISSIONS,
    UserRole.EDITOR: frozenset(EDITOR_PERMISSIONS),
    UserRole.ADMIN: frozenset(
        EDITOR_PERMISSIONS
        | {
            Permission.ADMIN_ACCESS,
            Permission.USERS_MANAGE,
            Permission.SOURCES_MANAGE,
            Permission.ADMIN_JOBS_MANAGE,
            Permission.AUDIT_VIEW,
            Permission.SYSTEM_VIEW,
            Permission.SYSTEM_SETTINGS_MANAGE,
            Permission.SEARCH_INDEX_VIEW,
            Permission.SEARCH_INDEX_MANAGE,
        }
    ),
}


def has_permission(role: UserRole | str, permission: Permission) -> bool:
    try:
        normalized = UserRole(role)
    except ValueError:
        return False
    return permission in PERMISSION_MATRIX[normalized]


def has_any_permission(role: UserRole | str, permissions: set[Permission]) -> bool:
    return any(has_permission(role, permission) for permission in permissions)


def has_all_permissions(role: UserRole | str, permissions: set[Permission]) -> bool:
    return all(has_permission(role, permission) for permission in permissions)

File: frontend/src/features/auth/permissions.ts
import type { Permission, PermissionMap, User, UserRole } from '../../types';

const userPermissions = [
  'SEARCH_USE',
  'RAG_USE',
  'PROFILE_MANAGE',
  'HISTORY_MANAGE',
  'SAVED_MANAGE',
] as const satisfies readonly Permission[];

const editorPermissions = [
  ...userPermissions,
  'EDITOR_ACCESS',
  'MANAGED_DOCUMENTS_VIEW',
  'DOCUMENT_METADATA_EDIT',
  'DOCUMENT_STATUS_CHANGE',
  'DOCUMENT_REINDEX',
  'EDITOR_JOBS_VIEW',
] as const satisfies readonly Permission[];

export const permissionMatrix: PermissionMap = {
  USER: userPermissions,
  EDITOR: editorPermissions,
  ADMIN: [
    ...editorPermissions,
    'ADMIN_ACCESS',
    'USERS_MANAGE',
    'SOURCES_MANAGE',
    'ADMIN_JOBS_MANAGE',
    'AUDIT_VIEW',
    'SYSTEM_VIEW',
    'SYSTEM_SETTINGS_MANAGE',
    'SEARCH_INDEX_VIEW',
    'SEARCH_INDEX_MANAGE',
  ],
};

export const roleLabels: Record<UserRole, string> = {
  USER: 'Пользователь',
  EDITOR: 'Редактор',
  ADMIN: 'Администратор',
};

export function hasPermission(
  subject: Pick<User, 'role' | 'accountStatus'> | UserRole | null | undefined,
  permission: Permission,
): boolean {
  if (!subject) return false;
  const role = typeof subject === 'string' ? subject : subject.role;
  if (typeof subject !== 'string' && subject.accountStatus !== 'ACTIVE') return false;
  return permissionMatrix[role].includes(permission);
}

export function hasAnyPermission(
  subject: Pick<User, 'role' | 'accountStatus'> | UserRole | null | undefined,
  permissions: readonly Permission[],
): boolean {
  return permissions.some((permission) => hasPermission(subject, permission));
}

export function hasAllPermissions(
  subject: Pick<User, 'role' | 'accountStatus'> | UserRole | null | undefined,
  permissions: readonly Permission[],
): boolean {
  return permissions.every((permission) => hasPermission(subject, permission));
}
